ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+# 18. Continuous Performance Testing (CPT) of Apps in Stonesoup
+
+Date: 2023-03-10
+
+## Status
+
+In consideration
+
+## Context
+
+In general, performance testing is just another form of testing that helps application teams to ensure there are no regressions in their code and that their application behaves as expected.
+
+The IntegrationTestScenario pipelines in StoneSoup are not suitable for full-blown performance and scale testing (that usually takes lots of time and involves human analysis), but are a good place for a quick and small-scale regression test.
+
+What makes performance testing different from functional testing is it is harder to decide if the test passed or failed as every performance test tests different aspects of the application and so it expects different performance, different metrics, different thresholds. Furthermore, even if the measured performance of the app under test did not change, there might be some significant resource usage by the application which we want to cause the test to be marked as failed.
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
```

Comment:
Can this phrasing be more precise? You don't have to go into implementation details, but an idea/proposal/future direction that doesn't include `thingy` :)

Reply:
OK, made it clearer I hope :)